src/leo_cache_tran.erl

```diff
@@ -0,0 +1,217 @@
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+stop() ->
+    gen_server:call(?MODULE, stop).
+
+%% @doc
+-spec(begin_tran(pid(), atom(), binary()) ->
+             ok | {error, in_process}).
+begin_tran(Pid, Tbl, Key) ->
+    gen_server:call(?MODULE, {begin_tran, Pid, Tbl, Key}, ?TRAN_TIMEOUT).
+
+
+%% @doc
+-spec(wait_tran(atom(), binary()) ->
+             {ok, any()} | {error, any()}).
+wait_tran(Tbl, Key) ->
+    wait_tran(Tbl, Key, ?TRAN_WAITTIME).
+
+%% @doc
+-spec(wait_tran(atom(), binary(), integer()) ->
+             {ok, any()} | {error, any()}).
+wait_tran(Tbl, Key, Waittime) ->
+    case catch gen_server:call(?MODULE, {wait_tran, Tbl, Key}, Waittime) of
+        {'EXIT', {timeout, _}} ->
+%%            gen_server:call(?MODULE, {unregister, Tbl, Key}, ?TRAN_TIMEOUT),
+            {error, timeout};
+        {'EXIT', Reason} ->
+%%            gen_server:call(?MODULE, {unregister, Tbl, Key}, ?TRAN_TIMEOUT),
+            {error, Reason};
+        Ret ->
+            Ret
+    end.
+
+%% @doc
+-spec(end_tran(atom(), binary())->
+            ok).
+end_tran(Tbl, Key) ->
+    gen_server:cast(?MODULE, {end_tran, Tbl, Key}).
+
+
+%%--------------------------------------------------------------------
+%% GEN_SERVER CALLBACKS
+%%--------------------------------------------------------------------
+%% Function: init(Args) -> {ok, State}          |
+%%                         {ok, State, Timeout} |
+%%                         ignore               |
+%%                         {stop, Reason}
+%% Description: Initiates the server
+init([]) ->
+    case ets:new(?PROCESSDB, [named_table, set, private]) of
+        ?PROCESSDB ->
+            ok;
+        _ ->
+            erlang:error({error, could_not_create_ets_table})
+    end,
+    case ets:new(?REPLYDB, [named_table, set, private]) of
+        ?REPLYDB ->
+            ok;
+        _ ->
+            erlang:error({error, could_not_create_ets_table})
+    end,
+    {ok, unused, ?TRAN_TIMEOUT}.
+
+handle_call(stop, _From, State) ->
+    {stop, normal, ok, State};
+
+%%handle_call({unregister, Tbl, Key}, From, State) ->
+%%    {Pid, _Ref} = From,
+%%    case ets:lookup(?REPLYDB, {Tbl, Key}) of
+%%        [{{Tbl, Key}, ReplyList}] ->
+%%            ReplyList2 = lists:filter(fun({Pid2, _Ref2}) ->
+%%                                              Pid2 =/= Pid
+%%                                      end, ReplyList),
+%%            true = ets:insert(?REPLYDB, {{Tbl, Key}, ReplyList2});
+%%        _ ->
+%%            void
+%%    end,
+%%    {reply, ok, State, ?TRAN_TIMEOUT};
+
+handle_call({begin_tran, Pid, Tbl, Key}, _From, State) ->
+    MonitorRef = erlang:monitor(process, Pid),
+    true = ets:insert(?PROCESSDB, {MonitorRef, {Tbl, Key}}), 
+    case ets:insert_new(?REPLYDB, {{Tbl, Key}, []}) of
+        true ->
+            {reply, ok, State, ?TRAN_TIMEOUT};
+        false ->
+            {reply, {error, in_process}, State, ?TRAN_TIMEOUT}
```

Comment:
Seems `erlang:demonitor` should be called if a caller coudn't be the owner process.